src/manifold/src/boolean_result.cpp

```diff
@@ -145,6 +165,81 @@ void AddNewEdgeVerts(
+#if MANIFOLD_PAR == 'T' && __has_include(<tbb/tbb.h>)
+  // parallelize operations, requires concurrent_map so we can only enable this
+  // with tbb
+  if (p1q2.size() > 128) {
+    // ideally we should have 1 mutex per key, but 64 is enough to avoid
+    // contention for most of the cases
+    std::array<std::mutex, 128> mutexes;
+    static tbb::affinity_partitioner ap;
+    tbb::parallel_for(
+        tbb::blocked_range<int>(0, p1q2.size(), 32),
+        [&](const tbb::blocked_range<int> &range) {
+          for (int i = range.begin(); i != range.end(); i++) {
+            const int edgeP = p1[i];
+            const int faceQ = q2[i];
+            const int vert = v12R[i];
+            const int inclusion = i12[i];
+            Halfedge halfedge = halfedgeP[edgeP];
+
+            {
+              size_t hash = std::hash<int>()(edgeP) % mutexes.size();
+              EdgePos edgePos = {vert, 0.0f, inclusion < 0};
+              auto &edgePosP = edgesP[edgeP];
+              // each EdgePos need to be locked independently,
+              // as pushing elements into a vector may reallocate and
+              // invalidate the pointer
+              mutexes[hash].lock();
+              for (int j = 0; j < glm::abs(inclusion); ++j) {
+                edgePosP.push_back(edgePos);
+                ++edgePos.vert;
+              }
+              mutexes[hash].unlock();
+            }
+
+            {
+              std::pair<int, int> key = {
+                  halfedgeP[halfedge.pairedHalfedge].face, faceQ};
+              if (!forward) std::swap(key.first, key.second);
+              size_t hash =
+                  std::hash<std::pair<int, int>>()(key) % mutexes.size();
+              EdgePos edgePos = {vert, 0.0f, inclusion < 0};
+              EdgePos edgePosRev = edgePos;
+              edgePosRev.isStart = !edgePos.isStart;
+              auto &edgePosRight = edgesNew[key];
+              mutexes[hash].lock();
+              for (int j = 0; j < glm::abs(inclusion); ++j) {
+                edgePosRight.push_back(forward ? edgePos : edgePosRev);
+                ++edgePos.vert;
+                ++edgePosRev.vert;
+              }
+              mutexes[hash].unlock();
+            }
+
+            {
+              std::pair<int, int> key = {halfedge.face, faceQ};
+              if (!forward) std::swap(key.first, key.second);
+              size_t hash =
+                  std::hash<std::pair<int, int>>()(key) % mutexes.size();
+              auto &edgePosLeft = edgesNew[key];
+              EdgePos edgePos = {vert, 0.0f, inclusion < 0};
+              EdgePos edgePosRev = edgePos;
+              edgePosRev.isStart = !edgePos.isStart;
+              mutexes[hash].lock();
+              for (int j = 0; j < glm::abs(inclusion); ++j) {
+                edgePosLeft.push_back(forward ? edgePosRev : edgePos);
+                ++edgePos.vert;
+                ++edgePosRev.vert;
+              }
+              mutexes[hash].unlock();
```

Comment:
There looks to be some duplicated code between these blocks - again functions would be nice so we don't have to keep changes in sync down the road.